compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -520,12 +527,164 @@ public void visit(BLangMatch matchStmt) {
     }
 
     private void analyzeStaticMatchPatterns(BLangMatch matchStmt) {
-        matchStmt.getStaticPatternClauses()
-                .stream()
-                .filter(pattern -> matchStmt.exprTypes
-                        .stream()
-                        .noneMatch(exprType -> this.types.isAssignable(exprType, pattern.literal.type)))
-                .forEach(pattern -> dlog.error(pattern.pos, DiagnosticCode.MATCH_STMT_UNMATCHED_PATTERN));
+        if (matchStmt.exprTypes.isEmpty()) {
+            return;
+        }
+        List<BLangExpression> matchedRecordPatterns = new ArrayList<>();
+        List<BLangExpression> matchedTuplePatterns = new ArrayList<>();
+        for (BLangMatchStmtStaticBindingPatternClause pattern : matchStmt.getStaticPatternClauses()) {
+            List<BType> matchedExpTypes = matchStmt.exprTypes
+                    .stream()
+                    .filter(exprType -> isValidMatchPattern(exprType, pattern.literal))
+                    .collect(Collectors.toList());
+            if (matchedExpTypes.isEmpty()) {
+                // log error if a pattern will not match to any of the expected types
+                dlog.error(pattern.pos, DiagnosticCode.MATCH_STMT_UNMATCHED_PATTERN);
+                continue;
+            }
+
+            if (pattern.getLiteral().type.tag == TypeTags.MAP) {
+                matchedRecordPatterns.add(pattern.getLiteral());
+                continue;
+            }
+
+            if (pattern.getLiteral().type.tag == TypeTags.TUPLE) {
+                matchedTuplePatterns.add(pattern.getLiteral());
+            }
+        }
+        analyzeUnreachableStaticPatterns(matchedRecordPatterns);
+        analyzeUnreachableStaticPatterns(matchedTuplePatterns);
+    }
+
+    private void analyzeUnreachableStaticPatterns(List<BLangExpression> matchedRecordPatterns) {
```

Comment:
lets rename the argument to be generic one as its is used for both tuple and record